### PR TITLE
Issue #6680: fix unexpected violation for extra empty line between package and import

### PR DESCRIPTION
**Pull request by gmshake**
Issue #6680 

Regression report: https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_3
https://romani.github.io/6680/diff/index.html

### CODE REVIEW

**Comment by gmshake** [comment]
@rnveach ping

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 1]
```diff
@@ -547,16 +555,21 @@ private void finishImportList() {
         String currentGroup = getImportGroup(firstImport.isStaticImport(),
                 firstImport.getImportFullPath());
         int currentGroupNumber = customImportOrderRules.indexOf(currentGroup);
+        ImportDetails previousImportObjectFromCurrentGroup = null;
         String previousImportFromCurrentGroup = null;
 
-        for (ImportDetails importObject : importToGroupList) {
+        for (final ImportDetails importObject : importToGroupList) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Please remove `final` from for loop variable.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 2]
```diff
@@ -756,21 +773,23 @@ private static int compareImports(String import1, String import2) {
...
+     * Counts empty lines between given.
+     * @param lineNoFrom
+     *        Line number of previous import.
+     * @param lineNoTo
      *        Line number of current import.
-     * @return count of empty lines before given.
+     * @return count of empty lines between given, exclusive, eg., (lineNoFrom, lineNoTo).
      */
-    private int getCountOfEmptyLinesBefore(int lineNo) {
+    private int getCountOfEmptyLinesBetween(int lineNoFrom, int lineNoTo) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: `lineNoFrom` => `fromLineNo`
`lineNoTo` => `toLineNo`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 3]
```diff
@@ -756,21 +773,23 @@ private static int compareImports(String import1, String import2) {
...
-        final String[] lines = getLines();
-        //  [lineNo - 2] is the number of the previous line
-        //  because the numbering starts from zero.
-        int lineBeforeIndex = lineNo - 2;
-        while (lineBeforeIndex >= 0
-                && CommonUtil.isBlank(lines[lineBeforeIndex])) {
-            lineBeforeIndex--;
-            result++;
+        if (lineNoFrom != lineNoTo) {
+            final String[] lines = Arrays.copyOfRange(getLines(), lineNoFrom, lineNoTo - 1);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Documentation says exclusive, but are you including `lineNoFrom` by not doing `lineNoFrom + 1`?

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Why do you need to copy `getLines` and not use it directly?

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties [hunk 4]
```diff
@@ -1,4 +1,5 @@
 custom.import.order=Import statement for ''{2}'' is in the wrong order. Should be in the ''{0}'' group, expecting group ''{1}'' on this line.
+custom.import.order.groups.separated.internally=Extra separation in import group before ''{0}''
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties]: Please make this a complete sentence. Other languages will have to be updated as well.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 5]
```diff
@@ -756,21 +773,23 @@ private static int compareImports(String import1, String import2) {
     }
 
     /**
-     * Counts empty lines before given.
-     * @param lineNo
+     * Counts empty lines between given.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: `between given` => `between given parameters`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 1 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Done!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 2 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Done!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: I've retested and confirmed that `Arrays.copyOfRange()` pass mutation test easily.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties (hunk 4 above)
Comment (gmshake) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties]: As the case, check line separator with import group, is the same with `ImportOrderCheck`,  the `custom.import.order.groups.separated.internally` was a duplicate of `import.groups.separated.internally` . Other languages were duplicated as well.

> Please make this a complete sentence.

I'm not sure how to describe the error message precisely, suggestions are welcome :)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 5 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Done!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: This isn't done.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: But it doesn't answer my original question. it seems like you are copying it for no reason.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 5 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: This was not done.

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Sorry about missed that :(

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Documented the parameter `fromLineNo` and `toLineNo` as one-based:
```java
    /**
     * Counts empty lines between given parameters.
     * @param fromLineNo
     *        One-based line number of previous import.
     * @param toLineNo
     *        One-based line number of current import.
     * @return count of empty lines between given parameters, exclusive,
     *        eg., (fromLineNo, toLineNo).
     */
    private int getCountOfEmptyLinesBetween(int fromLineNo, int toLineNo) {
        ...
    }
``` 

As `Arrays#copyOfRange(T[] original, int from, int to)`, the parameter `from` is inclusive but the third parameter `to` is exclusive, then `Arrays.copyOfRange(getLines(), fromLineNo, toLineNo - 1);` is effectively coping range [fromLineNo - 1 + 1, toLineNo - 1), or [fromLineNo - 1 + 1, toLineNo - 1 - 1].

Comments added above `Arrays.copyOfRang(..)`.

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: I'm not sure if this is proper to fake pitest, but the mutation test complains if using `getLines()` and looping with traditional index based `for` loop. The implementation looks the same, Ouch.

Traditional index based `for` loop that mutation test complains:
```java
    private int getCountOfEmptyLinesBetween(int fromLineNo, int toLineNo) {
        int result = 0;
        final String[] lines = getLines();

        for (int i = fromLineNo; i < toLineNo - 1; i++) {
            if (CommonUtil.isBlank(lines[i])) {
                result++;
            }
        }
        return result;
    }
```

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: If pitest fails for these lines, it means you don't have enough cases to fully satisfy them. I consider the copying of the arrays you did more faking out pitest.

What lines are failing pitest and what is error it is giving? You can upload the report to show it.

**Comment by gmshake** [comment]
@rnveach  See https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197/ for regression test report.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Uploaded pitest report. See https://gmshake.github.io/checkstyle_contribution/pit-reports-pr-7197-traditional-for-loop/201910311151/com.puppycrawl.tools.checkstyle.checks.imports/CustomImportOrderCheck.java.html  line 804

**Comment by romani** [comment]
@rnveach please finish review

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: https://gmshake.github.io/checkstyle_contribution/pit-reports-pr-7197-traditional-for-loop/201910311151/com.puppycrawl.tools.checkstyle.checks.imports/CustomImportOrderCheck.java.html#org.pitest.mutationtest.report.html.SourceFile@4a8e8f00_804
> 1. changed conditional boundary → SURVIVED

This means `i < toLineNo - 1` was changed to `i <= toLineNo - 1` and it caused no test failures.

If you need help on figuring how to kill this, you can always use regression.
See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-pitest-regression

You can try running a regression on using `<=` and see if any cases pop up. If none, you can change the condition to something like `i < toLineNo - 2` and so on until pitest is satisfied.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties (hunk 4 above)
Comment (rnveach) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties]: Lets leave it as it is then for now.

**Comment by rnveach** [comment]
@romani @gmshake 
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197/checkstyle/index.html#A2
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderNoPackage.java.html#L5

Isn't this issue only about violation between first import and package? Thats what i get form the first post of the issue. Regression between 2 imports shouldn't be changing, right?

**Comment by gmshake** [comment]
@rnveach 
As for https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderNoPackage.java.html#L5 , it is expected since a dedicated error message `MSG_SEPARATED_INTERNALLY` was introduced to make the error easier to understand by end users.

I'd like to file a separate PR with `MSG_SEPARATED_INTERNALLY` if it's preferred.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: @rnveach , is this discussion point resolved?

**Comment by romani** [comment]
sorry for long delay in reply ...

existing message: `custom.import.order.line.separator=''{0}'' should be separated from previous import group by one line.`
new message: `custom.import.order.groups.separated.internally=Extra separation in import group before ''{0}''`

@gmshake , please explain why new message is required ?
in issue we discussed and looks like come to agreement that there should be just no violation.  If you try to resolve some other issue, please create issue on it and lets make separate PR.

in this PR I expect just removal of extra violation, no new messages.

**Comment by gmshake** [comment]
> sorry for long delay in reply ...
> 
> existing message: `custom.import.order.line.separator=''{0}'' should be separated from previous import group by one line.`
> new message: `custom.import.order.groups.separated.internally=Extra separation in import group before ''{0}''`
> 
> @gmshake , please explain why new message is required ?
> in issue we discussed and looks like come to agreement that there should be just no violation. If you try to resolve some other issue, please create issue on it and lets make separate PR.
> 
> in this PR I expect just removal of extra violation, no new messages.

Ok, got it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: @romani No. `Arrays.copyOfRange` is still used.

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: I'd try as @rnveach suggested.

**Comment by gmshake** [comment]
See https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/ for regression test report.

**Comment by gmshake** [comment]
It looks there's something wrong with Travis CI build environment. The failing builds should use openjdk 11 and openjdk 12 but are actually using openjdk 13.0.1 which is not respect to the document, https://docs.travis-ci.com/user/reference/osx/#macos-version .

**Comment by romani** [comment]
Please rebase on latest master, I merged fix for this problem few minutes ago

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: @rnveach  Done at this moment. :)

**Comment by gmshake** [comment]
@romani Rebased to latest master.

**Comment by gmshake** [comment]
The CI looks good. :)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 3 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Ping @rnveach

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: @romani it is done

**Comment by romani** [comment]
I pushed extra commit to resolve Ncss issue, method is really complicated .... we need to fix it - done.

**Comment by romani** [comment]
@rnveach , please finalize review

**Comment by rnveach** [comment]
@romani please confirm https://github.com/checkstyle/checkstyle/pull/7197#issuecomment-550544664

**Comment by romani** [comment]
it is link to old report.
you mention - https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderNoPackage.java.html#L5

In latest report - https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/index.html there is no such violation.

attention to `diff_pr_7197` vs `diff_pr_7197_2`

**Comment by romani** [comment]
Do you mean that https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/index.html#A9 is not good violation ?

code:
```
19  package com.puppycrawl.tools.checkstyle.checks.whitespace.emptylineseparator;
20  
21  
22  
23  // violation
24  
25  
26  
27  import java.lang.Object; // violation
28  
29  import java.lang.Class; // no violation
30  
31  
32  // violation
33  import java.lang.Integer;
34  
35  // no violation
36  import java.lang.Long;
```

violation `warning | CustomImportOrder | 'java.lang.Integer' should be separated from previous import group by one line. | 33` definitely looks wrong, as group is the same.

**Comment by rnveach** [comment]
> attention to diff_pr_7197 vs diff_pr_7197_2

I didn't realize new report had difference in violations. I will re-review.

**Comment by rnveach** [comment]
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/index.html#A6
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationSeparatedStatements.java.html#L7
Also looks wrong as it is separated by 1 line.

**Comment by gmshake** [comment]
@romani 
> violation `warning | CustomImportOrder | 'java.lang.Integer' should be separated from previous import group by one line. | 33` definitely looks wrong, as group is the same.
> 
> fix for this is in dependent PR - https://github.com/checkstyle/checkstyle/pull/7198/files#diff-dd146a735b6f3d888dfcc709a6419c89R2
> 
> @gmshake , can we avoid such false-positives in this PR ?

See issue #7119 and #7278 , as they're definitely different kind of issue from this one.

**Comment by gmshake** [comment]
> https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/index.html#A6
> https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationSeparatedStatements.java.html#L7
> Also looks wrong as it is separated by 1 line.

Reasonable, it looks limited by the current logic of checking empty lines between import objects and the design of `com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck$ImportDetails`, as it has only one `lineNumber` but the import object span multiple lines.
https://github.com/checkstyle/checkstyle/blob/4b6ad27c3a5d27bcd746a7b7c69fcbc3ff25df41/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java#L865

@romani  @rnveach  At the first glance I would recommend we file another PR to fix such issue in the above situation, as it is rare enough. What do you think about that?

**Comment by gmshake** [comment]
After deep dig, I found that it's possible to fix the issue @rnveach pointed out

> https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationSeparatedStatements.java.html#L7

It requires slight modification of `com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck$ImportDetails` .

@romani  @rnveach  See the fix https://github.com/gmshake/checkstyle/pull/1  , and I'll merge it into this PR if you're satisfied with it.

**Comment by romani** [comment]
proposed PR to PR still not good.
In regression report there should be NO green lines at all.

**Comment by gmshake** [comment]
@romani Do you mean regressions such like 
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrder_OverlappingPatterns.java.html#L13 

and
 https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_2/checkstyle/xref/Users/zlei/Downloads/checkstyle_contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorWithComments.java.html#L33 ?

For this PR I change the logic counting  empty lines. The original one counts only empty lines before import, currently counts between import statements. I'll comment on the PR to show this change.

The logic can be reverted same as the original one and looks no values but prevent regressions as the above.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 6]
```diff
@@ -756,21 +819,23 @@ private static int compareImports(String import1, String import2) {
...
+     * Counts empty lines between given parameters.
+     * @param fromLineNo
+     *        One-based line number of previous import.
+     * @param toLineNo
+     *        One-based line number of current import.
+     * @return count of empty lines between given parameters, inclusive,
+     *        eg., [fromLineNo, toLineNo].
      */
-    private int getCountOfEmptyLinesBefore(int lineNo) {
+    private int getCountOfEmptyLinesBetween(int fromLineNo, int toLineNo) {
```
Comment (gmshake) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: The original logic counts only empty lines before given import statement. 
Current counts between import statements.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 7]
```diff
@@ -756,21 +819,23 @@ private static int compareImports(String import1, String import2) {
...
-        int lineBeforeIndex = lineNo - 2;
-        while (lineBeforeIndex >= 0
-                && CommonUtil.isBlank(lines[lineBeforeIndex])) {
-            lineBeforeIndex--;
-            result++;
+
+        // [fromLineNo - 1, toLineNo - 1 + 1)
+        for (int i = fromLineNo - 1; i < toLineNo; i++) {
+            if (CommonUtil.isBlank(lines[i])) {
+                result++;
```
Comment (gmshake) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Do NOT stop on comments between import statements.

**Comment by romani** [comment]
by fixing issue , you can not introduce new issue.
please find way to update logic of Check to have no new violations in diff report.
If you do not see how to do this, please let us know.

**Comment by gmshake** [comment]
@romani I've created an issue https://github.com/checkstyle/checkstyle/issues/7314 . To push on this issue we need to resolve https://github.com/checkstyle/checkstyle/issues/7314 first.

**Comment by gmshake** [comment]
Regression test report: https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_3

**Comment by gmshake** [comment]
Fixed misspellings and rebased on latest master.

**Comment by romani** [comment]
your changes were rebased to fix TC  and merged to first commit by https://roman-ivanov.blogspot.com/2019/05/shell-function-to-put-changes-in.html

**Comment by romani** [comment]
extended report looks good - https://romani.github.io/6680/diff/index.html

**Comment by gmshake** [comment]
@romani Following https://roman-ivanov.blogspot.com/2019/05/shell-function-to-put-changes-in.html the commit will lose original author, as it reset commits. 
You can rebase interactive to squash/fixup required commits into one and will preserve info as much as possible.
If you're a Macintosh user, there's an useful GUI tool https://github.com/git-up/GitUp to relief yourself. :)

**Comment by gmshake** [comment]
@romani Can you please restart the travis-ci build? As the spell check retrieved only partial english.words from checkstyle.sourceforge.io. A restart should be enough.

**Comment by romani** [comment]
All CIs are green now

**Comment by romani** [comment]
>You can rebase interactive to squash/fixup required commits into one and will preserve info as much as possible.

Interactive mode does not allow to squash commit in pre-previous.

**Comment by gmshake** [comment]
It’s a little off topic, but to squash commit with pre-previous you can swap the latest two commits and then do squashing.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 8]
```diff
@@ -756,21 +819,26 @@ private static int compareImports(String import1, String import2) {
...
-            lineBeforeIndex--;
-            result++;
+
+        //  "- 2" because the numbering starts from zero
+        for (int i = toLineNo - 2; i > fromLineNo - 2; i--) {
+            if (CommonUtil.isBlank(lines[i])) {
+                result++;
+            }
+            else {
+                // stop on none empty line to keep the behavior same as previous
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Should it be "stop on non-empty line..."?

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Good catch :)

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Fixed.

**Comment by gmshake** [comment]
Fixed typo and rebased on latest master.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: @gmshake , please share link to test that cover this catch.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 9]
```diff
@@ -756,21 +765,26 @@ private static int compareImports(String import1, String import2) {
...
-            lineBeforeIndex--;
-            result++;
+
+        //  "- 2" because the numbering starts from zero
+        for (int i = toLineNo - 2; i > fromLineNo - 2; i--) {
+            if (CommonUtil.isBlank(lines[i])) {
+                result++;
+            }
+            else {
+                // stop on non-empty line to keep the behavior same as previous
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: code is always inn current state .... there is no previous behavior.

if that is not good behavior .. please define `till issue https://.......` 
if that is expected, please remove "to keep the behavior same as previous"

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: There’s an ralated issue https://github.com/checkstyle/checkstyle/issues/7314 but has not been approved yet. Can you please take a look at it before I fix the comment?

Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Fixed :)

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: good for now.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: just comment is changed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 10]
```diff
@@ -593,6 +591,71 @@ private void finishImportList() {
...
+                firstImport.getImportFullPath());
+    }
+
+    /**
+     * Examine alphabetical order of imports.
+     *
+     * @param previousImport
+     *        previous import of current group.
+     * @param fullImportIdent
+     *        full import ident.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: > full import ident
> if previous and current import 

Javadoc definitions do not match up. This one should change to current import and variable should be renamed to as such.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 11]
```diff
@@ -593,6 +591,71 @@ private void finishImportList() {
...
+                && compareImports(fullImportIdent, previousImport) < 0;
+    }
+
+    /**
+     * Examine empty lines between groups.
+     *
+     * @param previousImportObject
+     *        previous import in current group.
+     * @param importObject
+     *        import group.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: > import group
> if current import NOT separated from previous import 

Javadoc definitions do not match up. This one should change to current import and variable should be renamed to as such.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 12]
```diff
@@ -593,6 +591,71 @@ private void finishImportList() {
...
+                     importObject.getLineNumber()) != 1;
+    }
+
+    /**
+     * Examine amount of empty lines.
+     *
+     * @param previousImportObject
+     *        previous import in current group.
+     * @param importObject
+     *        import group.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: same as others.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 13]
```diff
@@ -593,6 +591,71 @@ private void finishImportList() {
...
+     * @return
+     *        true, if current import separated from previous by more that one empty line.
+     */
+    private boolean isSeparatedByExtraLine(ImportDetails previousImportObject,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Javadoc specifies "empty" line but method name is extra line. Please synch up name.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 14]
```diff
@@ -756,21 +819,26 @@ private static int compareImports(String import1, String import2) {
     }
 
     /**
-     * Counts empty lines before given.
-     * @param lineNo
-     *        Line number of current import.
-     * @return count of empty lines before given.
+     * Counts empty lines between given parameters.
+     * @param fromLineNo
+     *        One-based line number of previous import.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: What does `one-based` mean? Normal numbers are 10-based. Are you saying this takes in binary number only?

**Comment by rnveach** [comment]
https://gmshake.github.io/checkstyle_contribution/checkstyle-tester/diff_pr_7197_3/
Regression looks good to me.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 14 above)
Comment (gmshake) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: The `one-based` was taken from https://github.com/checkstyle/checkstyle/blob/b2ad65928ec8ee3e76e5a1255a26d0ad85bcffbb/src/main/java/com/puppycrawl/tools/checkstyle/api/LineColumn.java#L30-L31 .

I think it is `one-based indexing` more precise.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Ok. Lets leave it for now.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 13 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 12 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 11 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java (hunk 10 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: done

**Comment by rnveach** [comment]
CI is failing.

**Comment by romani** [comment]
fixed, rebased ...

**Comment by romani** [comment]
@rnveach , PR is ready for review

**Comment by gmshake** [comment]
Thanks @romani  @rnveach  and @pbludov :)

**Pull request merged**